OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -39,6 +43,203 @@ static double UTOL = (double)SimTK::Eps*1e2;
 static double INTTOL = (double)SimTK::Eps*1e2;
 static int MAXITER = 20;
 static int NUM_SAMPLE_PTS = 100;
+
+//=============================================================================
+// PARAMETERS
+//=============================================================================
+
+/** Helper struct containing all parameters required to construct the
+ * SmoothSegmentedFunctionData.
+*/
+namespace OpenSim {
+struct SmoothSegmentedFunctionParameters {
+
+    SmoothSegmentedFunctionParameters(const SimTK::Matrix& mX,
+            const SimTK::Matrix& mY, double x0, double x1, double y0, double y1,
+            double dydx0, double dydx1, bool computeIntegral, bool intx0x1)
+            : _mX(mX), _mY(mY), _x0(x0), _x1(x1), _y0(y0), _y1(y1),
+              _dydx0(dydx0), _dydx1(dydx1), _computeIntegral(computeIntegral),
+              _intx0x1(intx0x1) {}
+
+    SimTK::Matrix _mX;
+    SimTK::Matrix _mY;
+
+    double _x0;
+    double _x1;
+    double _y0;
+    double _y1;
+    double _dydx0;
+    double _dydx1;
+    bool _computeIntegral;
+    bool _intx0x1;
+};
+
+bool operator==(
+    const SimTK::Matrix &lhs,
+    const SimTK::Matrix &rhs
+) {
+    if (lhs.nrow() != rhs.nrow() || lhs.ncol() != rhs.ncol()) {
+        return false;
+    }
+    for (int r = 0; r < lhs.nrow(); ++r)
+        for (int c = 0; c < lhs.ncol(); ++c)
+            if (lhs.row(r)[c] != rhs.row(r)[c]) return false;
+    return true;
+}
+
+bool operator==(
+    const SmoothSegmentedFunctionParameters &lhs,
+    const SmoothSegmentedFunctionParameters &rhs
+) {
+    return
+    lhs._x0 == rhs._x0 &&
+    lhs._x1 == rhs._x1 &&
+    lhs._y0 == rhs._y0 &&
+    lhs._y1 == rhs._y1 &&
+    lhs._dydx0 == rhs._dydx0 &&
+    lhs._dydx1 == rhs._dydx1 &&
+    lhs._computeIntegral == rhs._computeIntegral &&
+    lhs._intx0x1 == rhs._intx0x1 &&
+    lhs._mX == rhs._mX &&
+    lhs._mY == rhs._mY;
+}
+
+//=============================================================================
+// HASHING OF PARAMETERS
+//=============================================================================
+
+template<typename T>
+inline std::size_t HashCombine(std::size_t seed, T const &v) {
```

Comment:
`static` or anonymous namespace. Naming convention in OpenSim is `T const& v`, not `T const &v`. `std::size_t` is more correct, but other OpenSim code tends to use `size_t`.

Reply:
I think you mean `const T& v` right?

Reply:
I see `using namespace std` at the top, but also `std::string` rather than `string`. Is `size_t` special in this regard?

Reply:
Technically, yes, because `size_t` is a globally-defined data type from `<cstddef>`, but (if I understand correctly) it may not be defined, or differently defined, from `std::size_t`, which is the C++ standard's datatype for storing the size of an element in memory (for the given architecture, etc.)

Practically speaking, though, `static_assert(std::is_same_v<size_t, std::size_t>)` will pass in all codebases I've worked on up to now, and most codebases just use `size_t`.

As for `using namespace std;`, it means you can technically use `string` rather than `std::string`, but don't, because `using namespace std;` is bad practise (most codebases in the wild forbid it).